actionlint.cjs

```diff
@@ -8,54 +8,75 @@ require("./wasm_exec.js");
-  // executes the linter for all queued calls.
-  globalThis.actionlintInitialized = () => {
-    queued?.forEach((f) => f());
-    queued = globalThis.actionlintInitialized = undefined;
-  };
+  if (!(wasm.instance.exports.memory instanceof WebAssembly.Memory)) {
+    throw new Error("Could not get wasm memory");
+  }
+  const memory = wasm.instance.exports.memory;
+
+  if (!(wasm.instance.exports.WasmAlloc instanceof Function)) {
+    throw new Error("Could not get wasm alloc function");
+  }
+  const wasmAlloc = wasm.instance.exports.WasmAlloc;
+
+  if (!(wasm.instance.exports.WasmFree instanceof Function)) {
+    throw new Error("Could not get wasm free function");
+  }
+  const wasmFree = wasm.instance.exports.WasmFree;
 
-  loader(go).then((wasm) => {
-    // Do not await this promise, because it only resolves once the go main()
-    // function has exited. But we need the main function to stay alive to be
-    // able to call the `runActionlint` function.
-    go.run(wasm.instance);
-  });
+  if (!(wasm.instance.exports.RunActionlint instanceof Function)) {
+    throw new Error("Could not get wasm runActionLint function");
+  }
+  const runActionlint = wasm.instance.exports.RunActionlint;
 
   /**
-   * @param {string} src
+   * @param {string} input
    * @param {string} path
-   * @returns {Promise<LintResult[]>}
+   * @returns {LintResult[]}
    */
-  return async function runLint(src, path) {
-    // Return a promise, because we need to queue calls to `runLint()` while the
-    // wasm module is still loading and execute them once the wasm module is
-    //ready.
-    return new Promise((resolve, reject) => {
-      if (typeof runActionlint === "function") {
-        const [result, err] = runActionlint(src, path);
-        return err ? reject(err) : resolve(result);
-      }
-
-      if (!queued) {
-        queued = [];
-      }
-
-      queued.push(() => {
-        const [result, err] = runActionlint?.(src, path) ?? [
-          [],
-          new Error('"runActionlint" is not defined'),
-        ];
-        return err ? reject(err) : resolve(result);
-      });
-    });
+  return function runLint(input, path) {
+    const workflow = encoder.encode(input);
+    const filePath = encoder.encode(path);
+
+    const workflowPointer = wasmAlloc(workflow.byteLength);
+    new Uint8Array(memory.buffer).set(workflow, workflowPointer);
+
+    const filePathPointer = wasmAlloc(filePath.byteLength);
+    new Uint8Array(memory.buffer).set(filePath, filePathPointer);
+
+    const resultPointer = runActionlint(
+      workflowPointer,
+      workflow.byteLength,
+      workflow.byteLength,
+      filePathPointer,
+      filePath.byteLength,
+      filePath.byteLength
+    );
+
+    wasmFree(workflowPointer);
+    wasmFree(filePathPointer);
+
+    const result = new Uint8Array(memory.buffer).subarray(resultPointer);
+    const end = result.indexOf(0);
+
+    const string = decoder.decode(result.subarray(0, end));
```

Comment:
Since `result.indexOf(0)` is the length, should this be `result.subarray(1, end)`?

Reply:
Typically in TinyGo strings are represented as a tuple of `{ptr, len}` and slices are `{ptr, len, cap}` (https://tinygo.org/docs/concepts/compiler-internals/datatypes/), but since we can't return multiple values from wasm -> JS, we instead terminate our "string" / byte slice with a `NULL` byte (see: `main.go` L60) and return the pointer to the first element in our slice (see: `main.go` L81).
Thus, with `const result = new Uint8Array(memory.buffer).subarray(resultPointer);` we get an ArrayView from the first element of our byte slice and then search for the terminating null byte via `result.indexOf(0)`.

Reply:
My bad, I've read this as `result.at(0)` instead of `indexOf` which does search... :+1: